gap/fp/freeband.gi

```diff
@@ -420,6 +420,323 @@ function(S)
   fi;
 end);
 
+InstallMethod(EqualInFreeBand, "for two strings",
+[IsString, IsString],
+function(w1_in, w2_in)
+  if IsEmpty(w1_in) or IsEmpty(w2_in) then
+    return IsEmpty(w1_in) and IsEmpty(w2_in);
+    # Necessary since [] is recognised as a string yet the below
+    # doesn't work with it.
+  fi;
+  w1_in := List(w1_in, IntChar);
+  w2_in := List(w2_in, IntChar);
+  return EqualInFreeBand(w1_in, w2_in);
+end);
+
+# This is an implementation of the algorithm described in Jakub Radoszewski
+# and Wojciech Rytter's paper 'Efficient Testing of Equivalence of Words
+# in a Free Idempotent Semigroup⋆'
+
+InstallMethod(EqualInFreeBand, "for two lists of positive integers",
+[IsList, IsList],
+function(w1_in, w2_in)
+  local ListOfPosIntsToStandardListOfPosInts, Right, Left, LevelEdges,
+  RadixSort, StripFreeBandString, w1, w2, dollar, l1, l2, w, c, check,
+  rightk, leftk, edgecodes, rightm, leftm, i, k;
+
+  ListOfPosIntsToStandardListOfPosInts := function(list)
+    local L, distinct_chars, lookup, i;
+    if not IsList(list) then
+      ErrorNoReturn("expected a list as the argument");
+    fi;
+    for i in list do
+      if not IsPosInt(i) then
+        ErrorNoReturn("expected a list of positive integers as the argument");
+      fi;
+    od;
+    L := Length(list);
+    if L = 0 then
+      return [];
+    fi;
+
+    distinct_chars  := 1;
+    lookup          := [];
+    lookup[list[1]] := 1;
+    list[1]         := 1;
+
+    for i in [2 .. L] do
+      if IsBound(lookup[list[i]]) then
+        list[i] := lookup[list[i]];
+      else
+        distinct_chars  := distinct_chars + 1;
+        lookup[list[i]] := distinct_chars;
+        list[i]         := lookup[list[i]];
+      fi;
+    od;
+
+    return list;
+  end;
+
+  Right := function(w, k)
+    local max_char, right, i, j, content_size, multiplicity, length_w;
+    if not IsList(w) then
```

Comment:
Given that this function is not public (i.e. the only code that can use it belongs to this function), I don't think these error messages are necessary. What you're checking here is that the developers (not the users) of the Semigroups package are using this function correctly, so I'd suggest changing these errors to `Assert` instead.

Reply:
Postscript: this means that the checks will not be performed by GAP in some cases which might make this whole thing faster, without making it any less safe to use.